File: web-simulator/js/launcher.js
/**
 * TinyOLED Desktop — App Launcher (JavaScript Port)
 * 3×2 icon grid with paged navigation and selection animation.
 */

import { Font } from './font.js';

const CONTENT_Y = 10;
const CONTENT_H = 44;
const COLS = 3;
const ROWS = 2;
const CELL_W = Math.floor(128 / COLS);  // ~42
const CELL_H = Math.floor(CONTENT_H / ROWS);  // ~22

export class Launcher {
  constructor() {
    this._apps = [];
    this._cursor = 0;
    this._animOffset = 0;
  }

  register(name, label, icon, callback) {
    this._apps.push({ name, label, icon, callback });
  }

  moveUp() {
    if (this._cursor > 0) {
      this._cursor--;
      this._animOffset = 3;
    }
  }

  moveDown() {
    if (this._cursor < this._apps.length - 1) {
      this._cursor++;
      this._animOffset = -3;
    }
  }

  select() {
    if (this._cursor >= 0 && this._cursor < this._apps.length) {
      this._apps[this._cursor].callback();
    }
  }

  currentApp() {
    return this._apps[this._cursor] || null;
  }

  get appCount() {
    return this._apps.length;
  }

  draw(fb) {
    const perPage = COLS * ROWS;
    const visibleStart = Math.floor(this._cursor / perPage) * perPage;
    const visibleApps = this._apps.slice(visibleStart, visibleStart + perPage);

    for (let idx = 0; idx < visibleApps.length; idx++) {
      const app = visibleApps[idx];
      const col = idx % COLS;
      const row = Math.floor(idx / COLS);
      const cx = col * CELL_W;
      const cy = CONTENT_Y + row * CELL_H;
      const absIdx = visibleStart + idx;
      const selected = (absIdx === this._cursor);

      // Selected cell background
      if (selected) {
        const yOff = this._animOffset;
        fb.roundedRect(cx + 1, cy + yOff, CELL_W - 2, CELL_H - 2, 2, true, true);
      }

      // Icon (centered in cell, 8×8)
      const iconX = cx + Math.floor((CELL_W - 8) / 2);
      const iconY = cy + 2;
      fb.icon(app.icon, iconX, iconY, !selected);

      // Label (below icon)
      const label = app.label.slice(0, 5);
      const lw = Font.textWidth(label);
      const lx = cx + Math.floor((CELL_W - lw) / 2);
      fb.text(label, lx, iconY + 9, !selected);
    }

    // Decay animation
    if (this._animOffset > 0) this._animOffset--;
    else if (this._animOffset < 0) this._animOffset++;

    // Page indicators
    const totalPages = Math.ceil(this._apps.length / perPage);
    const curPage = Math.floor(this._cursor / perPage);
    if (totalPages > 1) {
      for (let p = 0; p < totalPages; p++) {
        const px = Math.floor(128 / 2) - totalPages * 3 + p * 6;
        if (p === curPage) {
          fb.rect(px, CONTENT_Y + CONTENT_H + 1, 4, 2, true, true);
        } else {
          fb.rect(px, CONTENT_Y + CONTENT_H + 1, 4, 2);
        }
      }
    }
  }
}

File: desktop/launcher.py
"""
TinyOLED Desktop — Uygulama Başlatıcı (App Launcher)
Status bar'ın altında 128×54 piksel alanda çalışır.

Düzen (3 sütun × 2 satır = 6 ikon):
  ┌──────────────────────────────────────────┐
  │  [CLK]  [SYS]  [WiFi]  [FILE] [SET] [PWR]│
  │  Saat   Sistem WiFi    Dosya  Ayar  Güç  │
  └──────────────────────────────────────────┘

Navigasyon:
  UP/DOWN → seçili ikonu kaydır
  SEL     → uygulamayı aç
"""

from typing import List, Callable, Optional
from core.framebuffer import Framebuffer
from core.font import Font

CONTENT_Y = 10   # status bar'ın altı
CONTENT_H = 44   # kullanılabilir yükseklik (notif yok iken)

COLS    = 3
ROWS    = 2
CELL_W  = 128 // COLS   # ~42 piksel
CELL_H  = CONTENT_H // ROWS  # ~22 piksel


class AppEntry:
    def __init__(self, name: str, label: str, icon: str, callback: Callable):
        self.name     = name
        self.label    = label
        self.icon     = icon      # 8×8 ikon adı (font.py içinde)
        self.callback = callback


class Launcher:
    """
    Ana ekran uygulama ızgarası.
    `select()` çağrıldığında ilgili app callback'i ateşlenir.
    """

    def __init__(self):
        self._apps: List[AppEntry] = []
        self._cursor: int = 0
        self._scroll: int = 0    # kaydırma (daha fazla uygulama için)
        self._anim_offset: int = 0   # seçim animasyonu

    def register(self, name: str, label: str, icon: str, callback: Callable):
        self._apps.append(AppEntry(name, label, icon, callback))

    def move_up(self):
        if self._cursor > 0:
            self._cursor -= 1
            self._anim_offset = 3

    def move_down(self):
        if self._cursor < len(self._apps) - 1:
            self._cursor += 1
            self._anim_offset = -3

    def select(self):
        if 0 <= self._cursor < len(self._apps):
            self._apps[self._cursor].callback()

    def current_app(self) -> Optional[AppEntry]:
        if 0 <= self._cursor < len(self._apps):
            return self._apps[self._cursor]
        return None

    def draw(self, fb: Framebuffer):
        """Uygulama ızgarasını çiz."""
        visible_start = (self._cursor // (COLS * ROWS)) * (COLS * ROWS)
        visible_apps  = self._apps[visible_start: visible_start + COLS * ROWS]

        for idx, app in enumerate(visible_apps):
            col = idx % COLS
            row = idx // COLS
            cx  = col * CELL_W
            cy  = CONTENT_Y + row * CELL_H
            abs_idx = visible_start + idx

            selected = (abs_idx == self._cursor)

            # Seçili hücre arka planı
            if selected:
                y_off = self._anim_offset
                fb.rounded_rect(cx + 1, cy + y_off, CELL_W - 2, CELL_H - 2,
                                r=2, on=True, fill=True)

            # İkon (hücre ortasında, 8×8)
            icon_x = cx + (CELL_W - 8) // 2
            icon_y = cy + 2
            fb.icon(app.icon, icon_x, icon_y, on=not selected)

            # Etiket (ikon altında, 5×7 font)
            label = app.label[:5]   # max 5 karakter (CELL_W sığdırmak için)
            lw    = Font.text_width(label)
            lx    = cx + (CELL_W - lw) // 2
            fb.text(label, lx, icon_y + 9, on=not selected)

        # Animasyon sönümle
        if self._anim_offset > 0:
            self._anim_offset -= 1
        elif self._anim_offset < 0:
            self._anim_offset += 1

        # Sayfa göstergesi (birden fazla sayfa varsa)
        total_pages = (len(self._apps) + COLS * ROWS - 1) // (COLS * ROWS)
        cur_page    = self._cursor // (COLS * ROWS)
        if total_pages > 1:
            for p in range(total_pages):
                px = 128 // 2 - total_pages * 3 + p * 6
                if p == cur_page:
                    fb.rect(px, CONTENT_Y + CONTENT_H + 1, 4, 2, fill=True)
                else:
                    fb.rect(px, CONTENT_Y + CONTENT_H + 1, 4, 2)
